keps/sig-node/20190408-evicting-pods-by-policy.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Evicting Pods by Policy
+authors:
+  - "@answer1991"
+owning-sig: sig-node
+reviewers:
+approvers:
+creation-date: 2019-04-07
+last-updated: 2019-04-07
+status: provisional
+
+---
+
+# Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [API](#api)
+    - [EvictReason Definition](#evictreason-definition)
+    - [EvictRequest Definition](#evictrequest-definition)
+    - [EvictPolicyType Definition](#evictpolicytype-definition)
+    - [EvictPolicy Definition](#evictpolicy-definition)
+  - [Workflow](#workflow)
+- [Implementation History](#implementation-history)
+
+# Summary
+
+`EvictingPodsByPolicy` feature is to enable users to define policies about how their pods are evicted when Nodes are unhealthy. This feature gives users more fine-grained control on evicting Pods, so as to make users application more robust during Pods migration when Nodes are unhealthy. 
+
+# Motivation
+
+For now, Kubernetes evict Pods running on the unhealthy Nodes by deleting them directly. We think this behavior is not a good practice in large scale Kubernetes cluster. For many applications, we may want to do some preparatory work, such as backup some meta data, before the Pods are evicted. What's more, we want to control the whole process of evicting Pods, including the speed of evicting Pods, the smarter desicion about when to delete the evicting Pods, enhancing the new Pods scheduling policy, etc.
+
+Kubernetes must provider a way to grant the control of Pod evicting process to users.
+
+### Goals
+
+- Define a CRD called `EvictRequest` which describes a Pod in reasons want be evicted by controller or kubelet.
+- Define a CRD called `EvictPolicy` to describe the evict policy of some Pods in some evicting reasons.
+
+- When controller or kubelet want to evict a Pod, they must create a `EvictRequest`. Only after the EvictRequest approved, the Pod can be evicted by controller or kubelet.
+- A controller called `approve-evict-request-controller ` is used to approve `EvictRequest` if the Pod evict policy is Default or none(the Pod can not find the related `EvictPolicy`).
+- Downstream Kubernetes developers can develop and delpoy their custom controllers to approve approve `EvictRequest` if the Pod evict policy is Custom.
+
+### Non-Goals
+
+- When and in which conditions to evicting Pods is NOT a goal.
+
+# Proposal
+
+### User Stories
+
+#### Story 1
+
+Before Pods are evicted, application developer want to backup some meta data of evicting Pods. What' more, if the evicting Pods are still running, we can backup some data of the evicting Pods to the new Pods, which may be helpful to speed up the new Pods ready to serve.
```

Comment:
if the pod is evicted due to resource constraints (i.e. memory, pid, disk), its possible these secondary actions may not be able to complete.  should these workloads run with higher priority or ensure they stay under their resource requests?

Reply:
Evicting Pods may influence the health of the service. For example, a Service has 3 Pods backends, if these 3 Pods was evicted as same time, then the Service is not healthy during eviction. If we can notify and transfer the eviction control to the workload controller(ie. replicaset controller) when we want to evict these Pods, then the workload controller will take a breath when evict these Pods one by one, and the Service will be healthy during eviction.

Reply:
If we can specified evict strategy to a replicaset resource, for example:

```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: foo
spec:
  replica: 3
  evictStrategy:
    maxUnavailable: 1
  ...
```

Then eviction-controller(node-lifecycle-controller) can notify replicaset-controller when we want to evict a Pod by adding an annotation on this Pod(i.e: evict: true).  

Then the replicaset-controller  start to evict Pods according the rs specified evictStrategy, replicaset-controller will ensure the rs resources exampled that only 1 Pod will be in eviction progress of this rs related Pods.